File: python/src/pi_mono/ai/providers/transform_messages.py
"""Message transformation utilities for cross-provider compatibility."""

from typing import Any

from pi_mono.ai.types import (
    AssistantMessage,
    ImageContent,
    Message,
    Model,
    TextContent,
    ToolCall,
)

NON_VISION_USER_IMAGE_PLACEHOLDER = "(image omitted: model does not support images)"
NON_VISION_TOOL_IMAGE_PLACEHOLDER = "(tool image omitted: model does not support images)"


def _replace_images_with_placeholder(
    content: list[TextContent | ImageContent], placeholder: str
) -> list[TextContent]:
    result: list[TextContent] = []
    previous_was_placeholder = False

    for block in content:
        if block["type"] == "image":
            if not previous_was_placeholder:
                result.append({"type": "text", "text": placeholder})
            previous_was_placeholder = True
            continue

        result.append(block)
        previous_was_placeholder = block.get("text", "") == placeholder

    return result


def _downgrade_unsupported_images(messages: list[Message], model: Model[Any]) -> list[Message]:
    if "image" in model.get("input", []):
        return messages

    downgraded = []
    for msg in messages:
        if msg["role"] == "user" and isinstance(msg.get("content"), list):
            downgraded.append(
                {
                    **msg,
                    "content": _replace_images_with_placeholder(
                        msg["content"], NON_VISION_USER_IMAGE_PLACEHOLDER
                    ),
                }
            )
        elif msg["role"] == "toolResult":
            downgraded.append(
                {
                    **msg,
                    "content": _replace_images_with_placeholder(
                        msg["content"], NON_VISION_TOOL_IMAGE_PLACEHOLDER
                    ),
                }
            )
        else:
            downgraded.append(msg)
    return downgraded


def _normalize_tool_call_id(id_: str, model: Model[Any], source: AssistantMessage) -> str:
    """Default normalization - override in specific providers."""
    return id_


def transform_messages(
    messages: list[Message],
    model: Model[Any],
    normalize_tool_call_id: Any = None,
) -> list[Message]:
    """
    Transform messages for cross-provider compatibility.

    Handles:
    - Unsupported image downgrade
    - Thinking block processing
    - Tool call ID normalization
    - Synthetic tool result insertion for orphaned calls
    """
    if normalize_tool_call_id is None:
        normalize_tool_call_id = _normalize_tool_call_id

    tool_call_id_map: dict[str, str] = {}
    image_aware_messages = _downgrade_unsupported_images(messages, model)

    # First pass: transform messages
    transformed: list[Message] = []
    for msg in image_aware_messages:
        # User messages pass through unchanged
        if msg["role"] == "user":
            transformed.append(msg)
            continue

        # Handle toolResult messages
        if msg["role"] == "toolResult":
            normalized_id = tool_call_id_map.get(msg["toolCallId"])
            if normalized_id and normalized_id != msg["toolCallId"]:
                transformed.append({**msg, "toolCallId": normalized_id})
            else:
                transformed.append(msg)
            continue

        # Assistant messages need transformation
        if msg["role"] == "assistant":
            assistant_msg: AssistantMessage = msg
            is_same_model = (
                assistant_msg.get("provider") == model.get("provider")
                and assistant_msg.get("api") == model.get("api")
                and assistant_msg.get("model") == model.get("id")
            )

            transformed_content = []
            for block in assistant_msg.get("content", []):
                if block["type"] == "thinking":
                    # Redacted thinking is opaque, only valid for same model
                    if block.get("redacted"):
                        if is_same_model:
                            transformed_content.append(block)
                        continue
                    # For same model: keep thinking blocks with signatures
                    if is_same_model and block.get("thinkingSignature"):
                        transformed_content.append(block)
                        continue
                    # Skip empty thinking blocks, convert others to plain text
                    if not block.get("thinking", "").strip():
                        continue
                    if is_same_model:
                        transformed_content.append(block)
                        continue
                    transformed_content.append({"type": "text", "text": block["thinking"]})
                    continue

                if block["type"] == "text":
                    if is_same_model:
                        transformed_content.append(block)
                    else:
                        transformed_content.append({"type": "text", "text": block["text"]})
                    continue

                if block["type"] == "toolCall":
                    tool_call: ToolCall = block
                    normalized_tool_call: ToolCall = tool_call.copy()

                    if not is_same_model and tool_call.get("thoughtSignature"):
                        normalized_tool_call = tool_call.copy()
                        normalized_tool_call.pop("thoughtSignature", None)

                    if not is_same_model:
                        normalized_id = normalize_tool_call_id(
                            tool_call["id"], model, assistant_msg
                        )
                        if normalized_id != tool_call["id"]:
                            tool_call_id_map[tool_call["id"]] = normalized_id
                            normalized_tool_call = normalized_tool_call.copy()
                            normalized_tool_call["id"] = normalized_id

                    transformed_content.append(normalized_tool_call)
                    continue

                transformed_content.append(block)

            transformed.append({**assistant_msg, "content": transformed_content})
            continue

        transformed.append(msg)

    # Second pass: insert synthetic tool results for orphaned calls
    result: list[Message] = []
    pending_tool_calls: list[ToolCall] = []
    existing_tool_result_ids: set[str] = set()

    def insert_synthetic_tool_results() -> None:
        nonlocal pending_tool_calls, existing_tool_result_ids
        if pending_tool_calls:
            for tc in pending_tool_calls:
                if tc["id"] not in existing_tool_result_ids:
                    result.append(
                        {
                            "role": "toolResult",
                            "toolCallId": tc["id"],
                            "toolName": tc["name"],
                            "content": [{"type": "text", "text": "No result provided"}],
                            "isError": True,
                            "timestamp": int(__import__("time").time() * 1000),
                        }
                    )
            pending_tool_calls = []
            existing_tool_result_ids = set()

    for msg in transformed:
        if msg["role"] == "assistant":
            insert_synthetic_tool_results()

            assistant_msg: AssistantMessage = msg
            if assistant_msg.get("stopReason") in ("error", "aborted"):
                continue

            tool_calls = [b for b in assistant_msg.get("content", []) if b["type"] == "toolCall"]
            if tool_calls:
                pending_tool_calls = tool_calls
                existing_tool_result_ids = set()

            result.append(msg)

        elif msg["role"] == "toolResult":
            existing_tool_result_ids.add(msg["toolCallId"])
            result.append(msg)

        elif msg["role"] == "user":
            insert_synthetic_tool_results()
            result.append(msg)

        else:
            result.append(msg)

    insert_synthetic_tool_results()
    return result

File: packages/ai/src/api/transform-messages.ts
import type {
	Api,
	AssistantMessage,
	ImageContent,
	Message,
	Model,
	TextContent,
	ToolCall,
	ToolResultMessage,
} from "../types.ts";

const NON_VISION_USER_IMAGE_PLACEHOLDER = "(image omitted: model does not support images)";
const NON_VISION_TOOL_IMAGE_PLACEHOLDER = "(tool image omitted: model does not support images)";

function replaceImagesWithPlaceholder(content: (TextContent | ImageContent)[], placeholder: string): TextContent[] {
	const result: TextContent[] = [];
	let previousWasPlaceholder = false;

	for (const block of content) {
		if (block.type === "image") {
			if (!previousWasPlaceholder) {
				result.push({ type: "text", text: placeholder });
			}
			previousWasPlaceholder = true;
			continue;
		}

		result.push(block);
		previousWasPlaceholder = block.text === placeholder;
	}

	return result;
}

function downgradeUnsupportedImages<TApi extends Api>(messages: Message[], model: Model<TApi>): Message[] {
	if (model.input.includes("image")) {
		return messages;
	}

	return messages.map((msg) => {
		if (msg.role === "user" && Array.isArray(msg.content)) {
			return {
				...msg,
				content: replaceImagesWithPlaceholder(msg.content, NON_VISION_USER_IMAGE_PLACEHOLDER),
			};
		}

		if (msg.role === "toolResult") {
			return {
				...msg,
				content: replaceImagesWithPlaceholder(msg.content, NON_VISION_TOOL_IMAGE_PLACEHOLDER),
			};
		}

		return msg;
	});
}

/**
 * Normalize tool call ID for cross-provider compatibility.
 * OpenAI Responses API generates IDs that are 450+ chars with special characters like `|`.
 * Anthropic APIs require IDs matching ^[a-zA-Z0-9_-]+$ (max 64 chars).
 */
export function transformMessages<TApi extends Api>(
	messages: Message[],
	model: Model<TApi>,
	normalizeToolCallId?: (id: string, model: Model<TApi>, source: AssistantMessage) => string,
): Message[] {
	// Build a map of original tool call IDs to normalized IDs
	const toolCallIdMap = new Map<string, string>();
	// Normalize null/undefined content from untyped callers (custom tools, hand-built
	// histories, old session files) so downstream code can rely on the type contract.
	const normalizedMessages = messages.map((msg) => (msg.content == null ? { ...msg, content: [] } : msg));
	const imageAwareMessages = downgradeUnsupportedImages(normalizedMessages, model);

	// First pass: transform messages (unsupported image downgrade, thinking blocks, tool call ID normalization)
	const transformed = imageAwareMessages.map((msg) => {
		// User messages pass through unchanged
		if (msg.role === "user") {
			return msg;
		}

		// Handle toolResult messages - normalize toolCallId if we have a mapping
		if (msg.role === "toolResult") {
			const normalizedId = toolCallIdMap.get(msg.toolCallId);
			if (normalizedId && normalizedId !== msg.toolCallId) {
				return { ...msg, toolCallId: normalizedId };
			}
			return msg;
		}

		// Assistant messages need transformation check
		if (msg.role === "assistant") {
			const assistantMsg = msg as AssistantMessage;
			const isSameModel =
				assistantMsg.provider === model.provider &&
				assistantMsg.api === model.api &&
				assistantMsg.model === model.id;

			const transformedContent = assistantMsg.content.flatMap((block) => {
				if (block.type === "thinking") {
					// Redacted thinking is opaque encrypted content, only valid for the same model.
					// Drop it for cross-model to avoid API errors.
					if (block.redacted) {
						return isSameModel ? block : [];
					}
					// For same model: keep thinking blocks with signatures (needed for replay)
					// even if the thinking text is empty (OpenAI encrypted reasoning)
					if (isSameModel && block.thinkingSignature) return block;
					// Skip empty thinking blocks, convert others to plain text
					if (!block.thinking || block.thinking.trim() === "") return [];
					if (isSameModel) return block;
					return {
						type: "text" as const,
						text: block.thinking,
					};
				}

				if (block.type === "text") {
					if (isSameModel) return block;
					return {
						type: "text" as const,
						text: block.text,
					};
				}

				if (block.type === "toolCall") {
					const toolCall = block as ToolCall;
					let normalizedToolCall: ToolCall = toolCall;

					if (!isSameModel && toolCall.thoughtSignature) {
						normalizedToolCall = { ...toolCall };
						delete (normalizedToolCall as { thoughtSignature?: string }).thoughtSignature;
					}

					if (!isSameModel && normalizeToolCallId) {
						const normalizedId = normalizeToolCallId(toolCall.id, model, assistantMsg);
						if (normalizedId !== toolCall.id) {
							toolCallIdMap.set(toolCall.id, normalizedId);
							normalizedToolCall = { ...normalizedToolCall, id: normalizedId };
						}
					}

					return normalizedToolCall;
				}

				return block;
			});

			return {
				...assistantMsg,
				content: transformedContent,
			};
		}
		return msg;
	});

	// Second pass: insert synthetic empty tool results for orphaned tool calls
	// This preserves thinking signatures and satisfies API requirements
	const result: Message[] = [];
	let pendingToolCalls: ToolCall[] = [];
	let existingToolResultIds = new Set<string>();
	// Tool results belonging to skipped errored/aborted assistant messages must be
	// dropped too, or providers that validate message order (e.g. Kimi) reject the
	// request: "messages with role 'tool' must be a response to a preceding message
	// with 'tool_calls'".
	const droppedToolCallIds = new Set<string>();
	// Synthetic "No result provided" results inserted below, tracked so they can be
	// dropped when the real result arrives later (see dedupe pass at the end).
	const syntheticToolCallIds = new Set<string>();
	const insertSyntheticToolResults = () => {
		if (pendingToolCalls.length > 0) {
			for (const tc of pendingToolCalls) {
				if (!existingToolResultIds.has(tc.id)) {
					result.push({
						role: "toolResult",
						toolCallId: tc.id,
						toolName: tc.name,
						content: [{ type: "text", text: "No result provided" }],
						isError: true,
						timestamp: Date.now(),
					} as ToolResultMessage);
					syntheticToolCallIds.add(tc.id);
				}
			}
			pendingToolCalls = [];
			existingToolResultIds = new Set();
		}
	};

	for (let i = 0; i < transformed.length; i++) {
		const msg = transformed[i];

		if (msg.role === "assistant") {
			// If we have pending orphaned tool calls from a previous assistant, insert synthetic results now
			insertSyntheticToolResults();

			// Skip errored/aborted assistant messages entirely.
			// These are incomplete turns that shouldn't be replayed:
			// - May have partial content (reasoning without message, incomplete tool calls)
			// - Replaying them can cause API errors (e.g., OpenAI "reasoning without following item")
			// - The model should retry from the last valid state
			const assistantMsg = msg as AssistantMessage;
			if (assistantMsg.stopReason === "error" || assistantMsg.stopReason === "aborted") {
				// Its tool results must not be replayed either: a `tool` message
				// without a preceding assistant `tool_calls` message is rejected by
				// providers that validate message order (e.g. Kimi).
				for (const block of assistantMsg.content) {
					if (block.type === "toolCall") droppedToolCallIds.add(block.id);
				}
				continue;
			}

			// Track tool calls from this assistant message
			const toolCalls = assistantMsg.content.filter((b) => b.type === "toolCall") as ToolCall[];
			if (toolCalls.length > 0) {
				pendingToolCalls = toolCalls;
				existingToolResultIds = new Set();
			}

			result.push(msg);
		} else if (msg.role === "toolResult") {
			if (droppedToolCallIds.has(msg.toolCallId)) continue;
			existingToolResultIds.add(msg.toolCallId);
			result.push(msg);
		} else if (msg.role === "user") {
			// User message interrupts tool flow - insert synthetic results for orphaned calls
			insertSyntheticToolResults();
			result.push(msg);
		} else {
			result.push(msg);
		}
	}

	// If the conversation ends with unresolved tool calls, synthesize results now.
	insertSyntheticToolResults();

	// Dedupe tool results: a synthetic "No result provided" may have been inserted
	// for a tool call whose real result arrives later (interrupted by a custom/user
	// message). Duplicate tool_call_ids make strict providers (e.g. Kimi) reject the
	// request with "Invalid request: tokenization failed".
	{
		const seen = new Set<string>();
		for (let i = result.length - 1; i >= 0; i--) {
			const msg = result[i];
			if (msg.role !== "toolResult") continue;
			if (seen.has(msg.toolCallId)) {
				if (syntheticToolCallIds.has(msg.toolCallId)) result.splice(i, 1);
				continue;
			}
			seen.add(msg.toolCallId);
		}
	}
	// Normalize tool-result placement. Providers that validate message order
	// (e.g. Kimi) reject requests where a `tool` message follows a `user` message
	// instead of the assistant message that issued the tool call. Session history
	// can interleave custom messages (e.g. background-task notifications) between
	// an assistant tool call and its result. Move each assistant's tool results to
	// immediately after it, before any interleaved user messages.
	for (let i = 0; i < result.length; i++) {
		const assistant = result[i];
		if (assistant.role !== "assistant") continue;
		const toolCallIds = new Set<string>(assistant.content.filter((b) => b.type === "toolCall").map((b) => b.id));
		if (toolCallIds.size === 0) continue;
		let k = i + 1;
		let interrupted = false;
		while (k < result.length) {
			const next = result[k];
			if (next.role === "toolResult" && toolCallIds.has(next.toolCallId)) {
				k++;
				continue;
			}
			if (next.role === "user") {
				interrupted = true;
				k++;
				continue;
			}
			break;
		}
		if (!interrupted) {
			i = k - 1;
			continue;
		}
		const segment = result.splice(i + 1, k - (i + 1));
		result.splice(
			i + 1,
			0,
			...segment.filter((m): m is ToolResultMessage => m.role === "toolResult"),
			...segment.filter((m) => m.role === "user"),
		);
		i = k - 1;
	}

	return result;
}
